File: ios/BasemapStyleConverter.swift
import ArcGIS

/// Converts the TypeScript `BasemapStyle` string-union values to the ArcGIS
/// `Basemap.Style` enum. Returns nil for unknown values so callers can raise a
/// stable `E_INVALID_ARGUMENT` rather than crashing. Pure, so it is unit-testable.
func basemapStyle(from value: String?) -> Basemap.Style? {
  switch value {
  case "arcGISStreets": return .arcGISStreets
  case "arcGISTopographic": return .arcGISTopographic
  case "arcGISNavigation": return .arcGISNavigation
  case "arcGISStreetsNight": return .arcGISStreetsNight
  case "arcGISDarkGray": return .arcGISDarkGray
  case "arcGISLightGray": return .arcGISLightGray
  case "arcGISImagery": return .arcGISImagery
  case "arcGISImageryStandard": return .arcGISImageryStandard
  case "arcGISOceans": return .arcGISOceans
  case "arcGISTerrain": return .arcGISTerrain
  default: return nil
  }
}

/// Derives a human-readable name from a service style path, e.g.
/// `arcgis/streets-night` → `Streets Night`. The service provides no display
/// name, so both platforms derive it identically for a consistent gallery.
func basemapDisplayName(from styleName: String) -> String {
  let last = styleName.split(separator: "/").last.map(String.init) ?? styleName
  return
    last
    .split(whereSeparator: { $0 == "-" || $0 == "_" })
    .map { $0.prefix(1).uppercased() + $0.dropFirst() }
    .joined(separator: " ")
}

/// Converts a `BasemapWorldview` string-union value to an ArcGIS `Worldview`.
func worldview(from value: String?) -> Worldview? {
  switch value {
  case "china": return .china()
  case "india": return .india()
  case "israel": return .israel()
  case "japan": return .japan()
  case "morocco": return .morocco()
  case "pakistan": return .pakistan()
  case "southKorea": return .southKorea()
  case "unitedArabEmirates": return .unitedArabEmirates()
  case "unitedStatesOfAmerica": return .unitedStatesOfAmerica()
  default: return nil
  }
}

/// Builds `BasemapStyleParameters` from the DTO, or nil when no tuning is set.
func makeStyleParameters(_ record: BasemapStyleParametersRecord?) -> BasemapStyleParameters? {
  guard let record, let view = worldview(from: record.worldview) else { return nil }
  let params = BasemapStyleParameters()
  params.worldview = view
  return params
}

File: android/src/main/java/expo/modules/arcgismapssdk/dto/BasemapStyleConverter.kt
package expo.modules.arcgismapssdk.dto

import com.arcgismaps.mapping.BasemapStyle
import com.arcgismaps.mapping.BasemapStyleParameters
import com.arcgismaps.mapping.Worldview

/**
 * Converts the TypeScript `BasemapStyle` string-union values to the ArcGIS
 * [BasemapStyle] enum. Returns null for unknown values so callers can raise a
 * stable `E_INVALID_ARGUMENT` rather than crashing.
 *
 * Kept as a pure function so it is unit-testable in isolation.
 */
fun basemapStyleFromString(value: String?): BasemapStyle? =
  when (value) {
    "arcGISStreets" -> BasemapStyle.ArcGISStreets
    "arcGISTopographic" -> BasemapStyle.ArcGISTopographic
    "arcGISNavigation" -> BasemapStyle.ArcGISNavigation
    "arcGISStreetsNight" -> BasemapStyle.ArcGISStreetsNight
    "arcGISDarkGray" -> BasemapStyle.ArcGISDarkGray
    "arcGISLightGray" -> BasemapStyle.ArcGISLightGray
    "arcGISImagery" -> BasemapStyle.ArcGISImagery
    "arcGISImageryStandard" -> BasemapStyle.ArcGISImageryStandard
    "arcGISOceans" -> BasemapStyle.ArcGISOceans
    "arcGISTerrain" -> BasemapStyle.ArcGISTerrain
    else -> null
  }

/** Converts a `BasemapWorldview` string-union value to an ArcGIS [Worldview]. */
fun worldviewFromString(value: String?): Worldview? =
  when (value) {
    "china" -> Worldview.china()
    "india" -> Worldview.india()
    "israel" -> Worldview.israel()
    "japan" -> Worldview.japan()
    "morocco" -> Worldview.morocco()
    "pakistan" -> Worldview.pakistan()
    "southKorea" -> Worldview.southKorea()
    "unitedArabEmirates" -> Worldview.unitedArabEmirates()
    "unitedStatesOfAmerica" -> Worldview.unitedStatesOfAmerica()
    else -> null
  }

/** Builds [BasemapStyleParameters] from the DTO, or null when no tuning is set. */
fun makeStyleParameters(record: BasemapStyleParametersRecord?): BasemapStyleParameters? {
  val view = worldviewFromString(record?.worldview) ?: return null
  return BasemapStyleParameters().apply { worldview = view }
}

/**
 * Derives a human-readable name from a service style path, e.g.
 * `arcgis/streets-night` -> `Streets Night`. The service provides no display
 * name, so both platforms derive it identically for a consistent gallery.
 */
fun basemapDisplayName(styleName: String): String =
  styleName
    .substringAfterLast('/')
    .split('-', '_')
    .filter { it.isNotEmpty() }
    .joinToString(" ") { it.replaceFirstChar(Char::uppercase) }
